graphql_compiler/compiler/ir_lowering_common/common.py

```diff
@@ -173,15 +181,22 @@ def visitor_fn(expression):
         else:
             return expression
 
-        expression_to_rewrite = None
+        expression_to_rewrite: Optional[BinaryComposition] = None
+        # N.B. cast statements occur in the last 2 elif statements because of the following:
+        # ```
+        # Incompatible types in assignment (expression has type "Expression", variable has type
+        # "Optional[BinaryComposition]")
+        # ```
+        # Given the logic at the time of writing this comment, we know that
+        # `right_is_binary_composition` says that right is BinaryComposition, not just Expression
         if expression.left == identity_literal and right_is_binary_composition:
             return expression.right
         elif expression.right == identity_literal and left_is_binary_composition:
             return expression.left
         elif expression.left == inverse_literal and right_is_binary_composition:
-            expression_to_rewrite = expression.right
+            expression_to_rewrite = cast(BinaryComposition, expression.right)
         elif expression.right == inverse_literal and left_is_binary_composition:
-            expression_to_rewrite = expression.left
+            expression_to_rewrite = cast(BinaryComposition, expression.left)
```

Comment:
This is fine, feel free to leave as is. I just wanted to give you an option to use an alternative approach, if you are dissatisfied with the current one, and would like to avoid the casts and the need for the comment.

The crux of the problem is that `mypy`'s static analysis is insufficiently powerful to see that `right_is_binary_composition == True` iff `isinstance(expression.right, BinaryComposition)`. This is true because:
- `right_is_binary_composition` is initialized with that precise statement on line 167 (in the new numbering), and
- `right_is_binary_composition` is never altered anywhere between that point and this point.

The easiest way to address this limitation of `mypy`'s is to make `left_binary_expression: Optional[BinaryComposition]` and `right_binary_expression: Optional[BinaryComposition]` variables, which are set to `expression.left` and `expression.right` if that side is a `BinaryComposition`, and `None` otherwise. Essentially, this is a type coercion operation, and in many languages, this operation is implemented via an `as` operator: `left_binary_expression = expression.left as BinaryComposition` would be the pseudo-Python way to express it here.

Then, the assignments on lines 197/199 don't need a cast, because they would read `expression_to_rewrite = left_binary_expression` which is valid no matter what, as the types of the two variables are definitionally the same.

Feel free to use your best judgment in deciding whether this alternative would be more or less readable than the current version.